text/0000-light-dom.md

```diff
@@ -0,0 +1,192 @@
+  [Styling is critical to web component reuse, but may prove difficult in practice](https://component.kitchen/blog/posts/styling-is-critical-to-web-component-reuse-but-may-prove-difficult-in-practice)
+
+- **Third party integrations**
+
+  Third party tools need to traverse the DOM, which breaks with Shadow DOM and the existing browser APIs (querySelector, ...). Note that the use of Light DOM fixes for Light DOM components, but not for native Shadow DOM ones if the page contains any.
+
+  - Analytics tools, Personalization platforms, Commerce tools like PriceSpider or Honey...
+
+- **Testing software**
+
+  Tools like Selenium, Cypress etc. face the same issues as third party tools when it comes to traversing the DOM.
+
+Introducing components rendering to the Light DOM opens new possibilities where applications can render most of their content in the Light DOM, but also keep individual widgets (e.g. `select` tag) and leaf components that render in the Shadow DOM. This approach combines the best of both worlds by solving the issues presented above and offering strong encapsulation when needed.
+![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")
+
+### Prior Art
+
+Most of the libraries designed to support Shadow DOM also propose a Light DOM option, with a variety of Shadow DOM features (slots, scoped styles, etc.). It includes:
+
+- [StencilJS](https://stenciljs.com/docs/styling#shadow-dom-in-stencil)
+- [LitElement](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)
+- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)
+
+## Detailed design
+
+### Selecting Light DOM vs Shadow DOM
+
+In raw JS, developer has the choice to use or not use shadow DOM for their custom element. For example,
+
+```js
+// uses Shadow DOM
+class MyShadowElement extends HTMLElement {
+  connectedCallback() {
+    this.attachShadow({ mode: "open" });
+    this.shadowRoot.innerHTML = "<p>My custom element that uses Shadow DOM";
+  }
+}
+
+// doesn't use Shadow DOM
+class MyLightElement extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = "<p>My element that does not use Shadow DOM";
+  }
+}
+```
+
+In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.
+
+```js
+import { MacroElement } from "lwc";
+
+export default class MyLightComponent extends MacroElement {
+  renderedCallback() {
+    this.querySelector("child-element"); // note the absence of this.template
+  }
+}
+```
+
+`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.
```

Comment:
I'm a bit concerned about exposing light DOM via `MacroElement`. I can break my concerns into two groups: 1) deviation from other WC frameworks, and 2) potential class proliferation.

## Deviation from other WC frameworks

Here is how other frameworks handle light DOM:

[Stencil](https://stenciljs.com/docs/host-element#shadow-dom)

```js
@Component({
  tag: 'my-component',
  shadow: false // or omitted
})
class MyComponent {}
```

[Fast](https://www.fast.design/docs/fast-element/working-with-shadow-dom/#shadow-dom-configuration)

```js
@customElement({
  name: 'my-component',
  template,
  shadowOptions: null // or `{ mode: 'closed' }` for closed shadow
})
class MyComponent extends FASTElement {}
```

[Lit](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)

```js
class MyComponent extends LitElement {
  createRenderRoot() {
    return this // or return `this.attachShadow(...)` here
  }
}
```

So we have 2 frameworks opting for decorators, and another opting for an overridable method.  LWC would be the only WC library using separate classes (unless I missed one).

## Potential class proliferation

In the future, we may want to support closed shadow roots. Would that be another class?

```js
import { ClosedShadowElement } from 'lwc'
```

There is also a proposal for "open stylable roots" (https://github.com/WICG/webcomponents/issues/909). Would this be another one?

```js
import { OpenStylableElement } from 'lwc'
```

In that proposal, there is even some talk of `attachShadow()` eventually looking like this:

```js
this.attachShadow({
  mode: 'open',
  styles: 'closed'
})
```

If the options bag for `attachShadow` continues to grow like this, then we could end up with a combinatorial explosion of classes to handle every possible case.

My personal preference would be for decorators or [mixins](https://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/), but mostly I'm just concerned about choosing classes here.

Reply:
I think the original reason we chose a different class was to emphasize that these are different _kinds_ of components because the programming model is different. I'm not sure if we still feel that is the case. 

I'm curious how the [createElement](https://github.com/salesforce/lwc/blob/ff6c28bc001b55cdee2a6f9442ce5e95a4f0e96a/packages/%40lwc/engine-dom/src/apis/create-element.ts#L89) API will change which currently takes the mode.

Reply:
@tedconn Ah my mistake, I didn't even know we supported closed shadow roots. 🙂  It seems that `createElement` would be the right place to put the options bag (if it becomes a big bag!).

So it looks like my concerns about class proliferation might not apply, since we only plan on having `MacroElement` for this one case (light DOM). But then why not have light DOM as an option in `createElement`?

Reply:
> But then why not have light DOM as an option in `createElement`

Because we need an API on the component, so component authors can toggle, not on the engine. Something like this would be nicer though, right?
```js
import { LightningElement } from 'lwc';

export default class LightDomElement extends LightningElement {

  static shadowOptions = {
    mode: null
  }

}
```

Reply:
We discussed in the past 2 options, using a static field or using a brand new base class. Until the decorator proposal goes back to stage 3, I don't think it is safe to introduce a brand new decorator.  

The main advantage of introducing a brand new class over static fields is around component inheritance. You can shadow (by inadvertence or intentionally) a base class static field. It means that you can turn a Shadow DOM component into a Light DOM one (and vice-versa) by extending from it. Such override can't be done when inheriting from a different base class.

In the following example, let say that we introduce a new `shadowDOM` static boolean to indicate whether the component rendered in the Light DOM or in the Shadow DOM. It is something we should discourage.

```
import { LigthningElement } from 'lwc';

class ShadowDOMComponent extends LightningElement {
  static shadowDOM = true;
}

class LigthDOMComponent extends ShadowDOMComponent {
  static shadowDOM = false;
}
``` ​

Now that you speak about class proliferation, I am more favorable to adding a signal via a static field.

Reply:
After discussion with the team, we landed on the static field rather than `MacroElement`, correct? AFAICR, we concluded:

- Any solution we land on (classes, field, method, decorators, etc.) needs to be resolved at runtime. In other words, developers can always do shenanigans such that you need to do a runtime evaluation to figure out whether the element is shadow or light. Since the LWC compiler to know at compile-time, we would just have to teach developers to use patterns that are friendly for compile-time evaluation.
- Decorators are not standard yet
- Classes may result in class proliferation
- Static fields seem to have nice ergonomics
- So we should go with static fields

Is this right? @pmdartus @abdulsattar